utils/db.ts

```diff
@@ -185,33 +185,42 @@ export async function getVotedItemIdsByUser(
     prefix: ["votes_by_users", userId],
   }, options);
   const voteItemIds = [];
-  for await (const res of iter) voteItemIds.push(res.key.at(-1));
+  for await (const res of iter) voteItemIds.push(res.key.at(-1) as string);
   return voteItemIds;
 }
 
 interface InitUser {
   id: string;
+  login: string;
+  avatarUrl: string;
   stripeCustomerId: string;
-  displayName: string;
+  sessionId: string;
 }
 
 export interface User extends InitUser {
   isSubscribed: boolean;
 }
 
-export async function createUser(initUser: InitUser) {
-  const usersKey = ["users", initUser.id];
-  const stripeCustomersKey = [
-    "user_ids_by_stripe_customer",
-    initUser.stripeCustomerId,
+export async function createUser(user: InitUser) {
+  const usersKey = ["users", user.id];
+  const usersByLoginKey = ["users_by_login", user.login];
+  const usersBySessionKey = ["users_by_session", user.sessionId];
+  const usersByStripeCustomerKey = [
+    "users_by_stripe_customer",
+    user.stripeCustomerId,
   ];
-  const user: User = { ...initUser, isSubscribed: false };
+
+  user = { ...user, isSubscribed: false } as User;
 
   const res = await kv.atomic()
     .check({ key: usersKey, versionstamp: null })
-    .check({ key: stripeCustomersKey, versionstamp: null })
+    .check({ key: usersByLoginKey, versionstamp: null })
+    .check({ key: usersBySessionKey, versionstamp: null })
+    .check({ key: usersByStripeCustomerKey, versionstamp: null })
     .set(usersKey, user)
-    .set(stripeCustomersKey, user.id)
+    .set(usersByLoginKey, user)
+    .set(usersBySessionKey, user)
+    .set(usersByStripeCustomerKey, user)
```

Comment:
I don't think it's good idea to save the same user in 4 different keys (The number could even grow when we add a new field to User model). Why not saving field-to-user_id mapping like the existing `user_ids_by_stripe_customer`?

Reply:
We previously used a `user_ids_by_stripe_customer` index that acts as a "join" between two indexes. However, the amount of coordination might be greater with that method. Saving and updating the same data across multiple indexes might be the way for KV databases. Other Deno KV implementations also follow this method (not that that's a real reason to follow).

Reply:
> However, the amount of coordination might be greater with that method.

Could you elaborate on this? Saving same data models in many keys looks more complex and causing more maintenance cost to me. Also it looks more error prone if someone forgot to update part of it in some update function.

> Other Deno KV implementations also follow this method

Can you give examples of this?

Reply:
> Could you elaborate on this? Saving same data models in many keys looks more complex and causing more maintenance cost to me. Also it looks more error prone if someone forgot to update part of it in some update function.

Say you have two indexes: `items` and `users`. If you need to get all items by a user, you'd probably have to have two intermediate indexes: `item_ids_to_user_ids` and `user_ids_to_item_ids`. This would mean you need to involve 2 indexes when getting all items by a user if you have the user ID. So, you have to find all items with that user ID using `kv.list()`, then get each item, likely using `kv.getMany()`, which is limited to 10 entries.

Compare this to the method used here. Using the user ID, you call `kv.list(["items_by_users", userId])` and you have all your items. To be clear, I'm not 100% sold on this method either, but so far, it seems most likely to be the best way to do it. As we discussed, it might be worth investing some time into a module that manages relational data within Deno KV better.

> Can you give examples of this?

- https://deno.com/manual@v1.33.3/runtime/kv/secondary_indexes
- https://github.com/denoland/tic-tac-toe
- https://github.com/hashrock/kv-sketchbook/blob/main/utils/db.ts

Reply:
hmm ok. Let's try multiple save strategy.